backend/imagekit/imagekit.go

```diff
@@ -0,0 +1,794 @@
+// return an error or upload it properly (rather than e.g. calling panic).
+//
+// May create the object even if it returns an error - if so
+// will return the object and the error, otherwise will return
+// nil and the error
+func (f *Fs) Put(ctx context.Context, in io.Reader, src fs.ObjectInfo, options ...fs.OpenOption) (fs.Object, error) {
+	return uploadFile(ctx, f, in, src.Remote(), options...)
+}
+
+// Mkdir makes the directory (container, bucket)
+//
+// Shouldn't return an error if it already exists
+func (f *Fs) Mkdir(ctx context.Context, dir string) error {
+	remote := path.Join(f.root, dir)
+
+	fs.Debugf(remote, "Creating directory")
+
+	parentFolderPath := f.EncodePath(remote[:strings.LastIndex(remote, "/")+1])
+	folderName := f.EncodePath(remote[strings.LastIndex(remote, "/")+1:])
+
+	_, e := f.ik.CreateFolder(ctx, client.CreateFolderParam{
+		ParentFolderPath: parentFolderPath,
+		FolderName:       folderName,
+	})
+
+	if e != nil {
+		return e
+	}
+
+	return nil
+}
+
+// Rmdir removes the directory (container, bucket) if empty
+//
+// Return an error if it doesn't exist or isn't empty
+func (f *Fs) Rmdir(ctx context.Context, dir string) error {
+	fs.Debugf(dir, "Removing directory")
+
+	entries, err := f.List(ctx, dir)
+
+	if err != nil {
+		return err
+	}
+
+	if len(entries) > 0 {
+		return errors.New("directory is not empty")
+	}
+
+	res, e := f.ik.DeleteFolder(ctx, client.DeleteFolderParam{
+		FolderPath: f.EncodePath(path.Join(f.root, dir)),
+	})
+
+	if res.StatusCode == http.StatusNotFound {
+		return fs.ErrorDirNotFound
+	}
+
+	if e != nil {
+		return e
+	}
+
+	return nil
+}
+
+// Purge deletes all the files and the container
+//
+// Optional interface: Only implement this if you have a way of
+// deleting all the files quicker than just running Remove() on the
+// result of List()
+func (f *Fs) Purge(ctx context.Context, dir string) error {
+	fs.Debugf(dir, "Purging directory")
+
+	remote := path.Join(f.root, dir)
+
+	res, e := f.ik.DeleteFolder(ctx, client.DeleteFolderParam{
+		FolderPath: f.EncodePath(remote),
+	})
+
+	if res.StatusCode == http.StatusNotFound {
+		return fs.ErrorDirNotFound
+	}
+
+	if e != nil {
+		return e
+	}
+
+	return nil
+}
+
+// PublicLink generates a public link to the remote path (usually readable by anyone)
+func (f *Fs) PublicLink(ctx context.Context, remote string, expire fs.Duration, unlink bool) (string, error) {
+
+	duration := time.Duration(math.Abs(float64(expire)))
+
+	expireSeconds := duration.Seconds()
+
+	fileRemote := path.Join(f.root, remote)
+
+	fs.Debugf(fileRemote, "Getting object")
+
+	folderPath := f.EncodePath(fileRemote[:strings.LastIndex(fileRemote, "/")+1])
+	fileName := f.EncodeFileName(fileRemote[strings.LastIndex(fileRemote, "/")+1:])
+
+	file := f.getFileByName(ctx, folderPath, fileName)
+
+	if file == nil {
+		return "", fs.ErrorObjectNotFound
+	}
+
+	url, err := f.ik.Url(client.UrlParam{
```

Comment:
Needs pacer?

Reply:
pacer would not be required for the Url call, it's generated at the client side only, no api call